src/libsync/propagateupload.cpp

```diff
@@ -197,6 +197,18 @@ void PropagateUploadFileCommon::start()
         return;
     }
 
+    // Check if we believe that the upload will fail due to remote quota limits
+    const quint64 quotaGuess = propagator()->_folderQuota.value(
+        QFileInfo(_item->_file).path(), std::numeric_limits<quint64>::max());
+    if (_item->_size > quotaGuess) {
+        // Necessary for blacklisting logic
+        _item->_httpErrorCode = 507;
+        _item->_errorMayBeBlacklisted = true;
+        emit propagator()->insufficientRemoteStorage();
+        done(SyncFileItem::BlacklistedError, tr("Upload of %1 exceeds the quota for the folder").arg(Utility::octetsToString(_item->_size)));
+        return;
+    }
```

Comment:
```
else  {
propagator()->_folderQuota[QFileInfo(_item->_file).path()] -= _item->_size;
}
```


(you probably want to use an  iterator to avoid multiple lookups, and only do it if there is a value there)